### PR TITLE
Constructor / prototype / frozen fixes

### PR DESCRIPTION
**Pull request by mpcarolin**
Wrote a helper function for deepClone that does the following:
- preserves the prototype in the clone
- clones all properties of an object instantiated with a constructor
- preserves the attributes of each property descriptor in the clone
- preserves the "isFrozen" state of the clone

I also wrote new tests for this. 

This was needed as it was breaking when trying to clone an object that had a constructor with arguments. Previously, deepClone would only call the constructor for the clone, but not pass in any of its arguments (since it did not know them). Now, it creates a new object with the source's prototype and properties.

### CODE REVIEW

File: src/scripts/collection.js [hunk 1]
```diff
@@ -202,9 +201,14 @@ export const deepClone = (obj, hash = new WeakMap()) => {
...
       ? new RegExp(obj.source, obj.flags)
-      : obj.constructor 
-        ? new obj.constructor()
-        : Object.create(null)
+      : (!obj.constructor)
+        ? Object.create(null)
+        : null
+  
+  // if result is null, object has a constructor and wasn't an instance of Date nor RegExp
+  if (result === null) {
```
Comment (lancetipton) [inline src/scripts/collection.js]: For single lines like this I remove the `{}`. Makes it less verbose like you do down on line 232

File: src/scripts/collection.js [hunk 2]
```diff
@@ -217,4 +221,27 @@ export const deepClone = (obj, hash = new WeakMap()) => {
...
+  
+  const sourceDescriptors = Object.getOwnPropertyDescriptors(objectWithPrototype)
+
+  for (const [key, descriptor] of Object.entries(sourceDescriptors)) {
+    sourceDescriptors[key].value = deepClone(descriptor.value)
+  }
+
+  const clone = Object.create(prototype, sourceDescriptors)
+
+  return Object.isFrozen(objectWithPrototype)
```
Comment (lancetipton) [inline src/scripts/collection.js]: We should add  a check for `Object.seal` with `Object.isSealed`

File: src/scripts/__tests__/collection.js [hunk 3]
```diff
@@ -240,9 +240,31 @@ describe('/collection', () => {
       expect(clone.das[0].bar === org.das[0].bar).toEqual(false)
       expect(clone.das[0].bar[0]).toEqual('foo')
       expect(clone.das[0].bar[1]).toEqual('baz')
+    })
 
+    it('should make a frozen clone if the source is frozen', () => {
```
Comment (lancetipton) [inline src/scripts/__tests__/collection.js]: When you add the `seal` check, you could just dup this test, and change Freeze to Seal

**Comment by mpcarolin** [comment]
Updated

**Comment by mpcarolin** [comment]
Found another issue that I'm fixing now. It clones arrays but it just makes them into Objects whose prototype is array. Not quite the same... (it will fail Array.isArray(x))

**Pull request merged**